### PR TITLE
wallet: Don't generate keys for wallets with private keys disabled during upgradewallet

### LINKED ISSUE #23610: Calling `upgradewallet` on a legacy wallet causes Bitcoin Core to crash
I have 2 legacy wallets and on both if I run `upgradewallet` it will cause bitcoin core to crash. This should just return me an error string

Currenl result from `getwalletinfo`:
```
{
  "walletname": "Wallet",
  "walletversion": 169900,
  "format": "bdb",
  "balance": 0.00000000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 0.00000000,
  "txcount": 80,
  "keypoololdest": 1613078825,
  "keypoolsize": 941,
  "keypoolsize_hd_internal": 941,
  "paytxfee": 0.00000000,
  "private_keys_enabled": false,
  "avoid_reuse": true,
  "scanning": false,
  "descriptors": false
}
```

### PR DESCRIPTION
**Pull request by achow101**
When we're upgrading a wallet, we shouldn't be trying to generate new keys for wallets where private keys are disabled.

Fixes #23610

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

File: src/wallet/scriptpubkeyman.cpp [hunk 1]
```diff
@@ -469,6 +469,12 @@ bool LegacyScriptPubKeyMan::CanGetAddresses(bool internal) const
 bool LegacyScriptPubKeyMan::Upgrade(int prev_version, int new_version, bilingual_str& error)
 {
     LOCK(cs_KeyStore);
+
+    if (m_storage.IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```
Comment (w0xlt) [inline src/wallet/scriptpubkeyman.cpp:473]: nit: If the user is expecting the wallet upgrade, shouldn't this condition return false and inform the reason in `error`?

```suggestion
    if (m_storage.IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
            error = _("Private keys disabled");
            return false;
    }
```

**Review by w0xlt — APPROVED** [review]
tACK 34afde0 on Ubuntu 21.10

File: src/wallet/scriptpubkeyman.cpp (hunk 1 above)
Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp:473]: No, the wallet is still being upgraded - the version number will change, and in the future (perhaps in `CWallet`), there may actually be an upgrade. It is fine for an upgrade to do nothing - that is what we do for descriptor wallets.

File: test/functional/wallet_upgradewallet.py [hunk 2]
```diff
@@ -345,5 +345,11 @@ def copy_split_hd():
             desc_wallet = self.nodes[0].get_wallet_rpc("desc_upgrade")
             self.test_upgradewallet(desc_wallet, previous_version=169900, expected_version=169900)
 
+        if self.is_bdb_compiled():
```
Comment (instagibbs) [inline test/functional/wallet_upgradewallet.py:348]: can we also test descriptor+disable_private_keys while we're here?

Comment (achow101) [reply test/functional/wallet_upgradewallet.py:348]: Done

**Review by benthecarman — APPROVED** [review]
tACK c7376cc8d728f3a7c40f79bd57e7cef685def723 this fixed the issue for me

**Comment by fanquake** [comment]
@w0xlt want to take another look here?

**Comment by laanwj** [comment]
Code review ACK c7376cc8d728f3a7c40f79bd57e7cef685def723

**Pull request merged**